File: matlab/Data_layer_Polygons.m
%% MATLAB code for coordinate location tracking and polygon assignment
% Step 2 of the DFO Calanus biomass processing workflow.
% Assigns each depth-integrated grid point (from DFO_data_process.R output)
% to a CINAR region polygon and an EcoMon survey stratum.
%
% Inputs:  *_processed.csv files (one per month/year) from input_dir
% Outputs: *_processed_polygons.csv files with CINAR_poly and EcoMon_poly columns
%
% CINAR polygon ID mapping:
%   1 = WSS (Western Scotian Shelf)
%   2 = EGOM (Eastern Gulf of Maine)
%   3 = JB (Jordan Basin)
%   4 = Browns (Browns Bank)
%   5 = Halifax (Eastern Scotian Shelf)
%   6 = GeorgesNEC (Georges Basin and NE Channel)
%   7 = GMB150 (Grand Manan Basin, 150 m isobath)
%   8 = BOF (Bay of Fundy)
%   0 = Unassigned
%
% Assignment priority (highest first): GMB150 > JB > GeorgesNEC > BOF > WSS > EGOM > Browns > Halifax
% A point is assigned to the first polygon in this sequence that contains it.
%
% NOTE: For a faster, fully-R implementation see Data_layer_Polygons.R,
% which replicates this logic using sf::st_difference() + sf::st_join().

clc; clear; close all;

%% Paths
% Repository root — resolved from the location of this script file.
work_dir   = fileparts(mfilename('fullpath'));
input_dir  = fullfile(work_dir, 'processed');
output_dir = fullfile(work_dir, 'polygons');

% sp_proj.m lives in work_dir; add it to path
addpath(work_dir);

if ~exist(output_dir, 'dir')
    mkdir(output_dir);
end

%% Load polygon coordinate files
% All files are in decimal degrees (lon, lat).
% GMB_150 and GeorgesNEC use the _latlon.csv versions directly.
% The _sp.csv (state plane) files are not used here.
fprintf('Loading polygon coordinate files...\n');

Browns_line            = load(fullfile(work_dir, 'Browns_line.txt'));
Halifax_line           = load(fullfile(work_dir, 'Halifax_line.txt'));
JB_deep_latlon         = readmatrix(fullfile(work_dir, 'JB_deep_latlon.csv'));
GeorgesNEC_deep_latlon = readmatrix(fullfile(work_dir, 'GeorgesNEC_deep_latlon.csv'));
GMB_150_latlon         = readmatrix(fullfile(work_dir, 'GMB_150_latlon.csv'));
BOF_latlon             = readmatrix(fullfile(work_dir, 'BOF_latlon.csv'));
WSS_broad              = readmatrix(fullfile(work_dir, 'WSS_broad.csv'));
EGOM_broad             = readmatrix(fullfile(work_dir, 'EGOM_broad.csv'));

%% Prepare EcoMon shape objects
fprintf('Loading EcoMon polygons from EMstrata_v4.mat...\n');
load(fullfile(work_dir, 'EMstrata_v4.mat'));

% Build polyshape objects for EcoMon strata indices 14-47.
% These index values match the NES stratum numbering in EMstrata_v4.
EcoMonshape = polyshape.empty;
for j = 14:47
    x = EMstrata_v4(j).x;
    y = EMstrata_v4(j).y;
    EcoMonshape(j) = polyshape(x, y);
end

%% Prepare CINAR shape objects
fprintf('Creating CINAR polygon shapes...\n');

% Extract coordinates (column 1 = lon, column 2 = lat throughout)
JB_x         = JB_deep_latlon(:, 1);
JB_y         = JB_deep_latlon(:, 2);
Browns_x     = Browns_line(:, 1);
Browns_y     = Browns_line(:, 2);
Halifax_x    = Halifax_line(:, 1);
Halifax_y    = Halifax_line(:, 2);
BOF_x        = BOF_latlon(:, 1);
BOF_y        = BOF_latlon(:, 2);
WSS_x        = WSS_broad(:, 1);
WSS_y        = WSS_broad(:, 2);
EGOM_x       = EGOM_broad(:, 1);
EGOM_y       = EGOM_broad(:, 2);
GMB150_x     = GMB_150_latlon(:, 1);
GMB150_y     = GMB_150_latlon(:, 2);
GeorgesNEC_x = GeorgesNEC_deep_latlon(:, 1);
GeorgesNEC_y = GeorgesNEC_deep_latlon(:, 2);

% Create CINAR polyshape objects; index = CINAR_poly ID
shape(1) = polyshape(WSS_x,        WSS_y);        % 1 = WSS
shape(2) = polyshape(EGOM_x,       EGOM_y);       % 2 = EGOM
shape(3) = polyshape(JB_x,         JB_y);         % 3 = JB
shape(4) = polyshape(Browns_x,     Browns_y);     % 4 = Browns
shape(5) = polyshape(Halifax_x,    Halifax_y);    % 5 = Halifax
shape(6) = polyshape(GeorgesNEC_x, GeorgesNEC_y); % 6 = GeorgesNEC
shape(7) = polyshape(GMB150_x,     GMB150_y);     % 7 = GMB150
shape(8) = polyshape(BOF_x,        BOF_y);        % 8 = BOF

% Polygon names indexed by CINAR_poly ID (used in summary output)
polygon_names = {'WSS', 'EGOM', 'JB', 'Browns', 'Halifax', 'GeorgesNEC', 'GMB150', 'BOF'};

%% Get list of processed CSV files
csv_files = dir(fullfile(input_dir, '*_processed.csv'));
fprintf('Found %d processed CSV files to analyze\n', length(csv_files));

%% Process each CSV file
for file_idx = 1:length(csv_files)
    filename = csv_files(file_idx).name;
    fprintf('\n%s\n', repmat('=', 1, 60));
    fprintf('Processing file: %s\n', filename);

    filepath = fullfile(input_dir, filename);

    try
        data_table_full = readtable(filepath);
        fprintf('Loaded data: %d rows x %d columns\n', height(data_table_full), width(data_table_full));

        % Check required columns
        required_vars = {'X', 'Y', 'REGION'};
        if ~all(ismember(required_vars, data_table_full.Properties.VariableNames))
            fprintf('Warning: Required columns (X, Y, REGION) not found in %s. Skipping.\n', filename);
            continue;
        end

        % Filter to the six DFO source regions used in the analysis
        desired_regions = {'CCB', 'Fundy', 'GB', 'GOM', 'SNE', 'SS'};
        data_table = data_table_full(ismember(data_table_full.REGION, desired_regions), :);
        fprintf('After regional filter: %d rows remaining\n', height(data_table));

        if height(data_table) == 0
            fprintf('Warning: No data remaining after regional filtering. Skipping.\n');
            continue;
        end

        % Initialise assignment columns (0 = unassigned)
        data_table.CINAR_poly  = zeros(height(data_table), 1);
        data_table.EcoMon_poly = zeros(height(data_table), 1);

        %% CINAR polygon assignment
        fprintf('Assigning CINAR polygons...\n');
        for i = 1:height(data_table)
            if mod(i, 1000) == 0
                fprintf('  Processing row %d of %d\n', i, height(data_table));
            end

            pt_x = data_table.X(i);
            pt_y = data_table.Y(i);

            if isnan(pt_x) || isnan(pt_y)
                continue;
            end

            % Priority order: GMB150(7), JB(3), GeorgesNEC(6), BOF(8),
            %                 WSS(1), EGOM(2), Browns(4), Halifax(5)
            for shape_idx = [7, 3, 6, 8, 1, 2, 4, 5]
                if isinterior(shape(shape_idx), pt_x, pt_y)
                    data_table.CINAR_poly(i) = shape_idx;
                    break;
                end
            end
        end

        %% EcoMon polygon assignment
        fprintf('Assigning EcoMon polygons...\n');
        for i = 1:height(data_table)
            if mod(i, 1000) == 0
                fprintf('  Processing row %d of %d\n', i, height(data_table));
            end

            pt_x = data_table.X(i);
            pt_y = data_table.Y(i);

            if isnan(pt_x) || isnan(pt_y)
                continue;
            end

            for j = 14:47
                if ~isempty(EcoMonshape(j).Vertices) && isinterior(EcoMonshape(j), pt_x, pt_y)
                    data_table.EcoMon_poly(i) = j;
                    break;
                end
            end
        end

        %% Save output
        [~, base_name, ~] = fileparts(filename);
        output_filename   = [base_name, '_polygons.csv'];
        output_filepath   = fullfile(output_dir, output_filename);
        writetable(data_table, output_filepath);

        %% Print assignment summary
        cinar_counts       = histcounts(data_table.CINAR_poly, 0.5:8.5);
        n_unassigned_cinar = sum(data_table.CINAR_poly == 0);
        ecomon_assigned    = sum(data_table.EcoMon_poly > 0);

        fprintf('CINAR polygon assignments:\n');
        for i = 1:8
            if cinar_counts(i) > 0
                fprintf('  %s (%d): %d points\n', polygon_names{i}, i, cinar_counts(i));
            end
        end
        fprintf('  Unassigned: %d points\n', n_unassigned_cinar);
        fprintf('EcoMon assignments: %d points assigned\n', ecomon_assigned);
        fprintf('Saved: %s\n', output_filename);

    catch ME
        fprintf('Error processing file %s: %s\n', filename, ME.message);
    end
end

fprintf('\n%s\n', repmat('=', 1, 60));
fprintf('Processing complete!\n');
fprintf('Output files saved to: %s\n', output_dir);

File: R/Data_layer_Polygons.R
## Data_layer_Polygons.R
## Step 2 of the DFO Calanus biomass workflow.
## Assigns each depth-integrated grid point to a CINAR region polygon and an
## EcoMon survey stratum using point-in-polygon operations (sf package).
##
## Required packages: dplyr, sf
## Install if needed: install.packages(c("dplyr", "sf"))
##
## Input files (repo root):
##   poly_*.csv          — pre-clipped CINAR polygon boundaries (from export_for_R.m)
##   EMstrata_v4_coords.csv — EcoMon stratum polygon coordinates (from EMstrata_v4.mat)
##   ne_strata_cache.rds — cached sf object built on first run; delete to rebuild
##
## CINAR polygon ID mapping (CINAR_poly column):
##   1 = WSS (Western Scotian Shelf)
##   2 = EGOM (Eastern Gulf of Maine)
##   3 = JB (Jordan Basin)
##   4 = Browns (Browns Bank)
##   5 = Halifax (Eastern Scotian Shelf)
##   6 = GeorgesNEC (Georges Basin and NE Channel)
##   7 = GMB150 (Grand Manan Basin, 150 m isobath)
##   8 = BOF (Bay of Fundy)
##   0 = Unassigned

library(dplyr)
library(sf)

# Use planar (GEOS) geometry instead of spherical (S2).
# S2 rejects polygon files with duplicate consecutive vertices; GEOS tolerates
# them. Planar geometry is appropriate for this regional northwest Atlantic domain.
sf_use_s2(FALSE)

# ===========================================================================
# Configuration
# ===========================================================================

# Repository root — set automatically from the current working directory.
# Open the .Rproj file (or setwd() to the repo root) before sourcing.
work_dir   <- getwd()
input_dir  <- file.path(work_dir, "processed")
output_dir <- file.path(work_dir, "polygons")

if (!dir.exists(output_dir)) dir.create(output_dir, recursive = TRUE)

# Six DFO source regions retained in the analysis
desired_regions <- c("CCB", "Fundy", "GB", "GOM", "SNE", "SS")

# ===========================================================================
# Helper: read a polygon coordinate file into a closed-ring matrix
# ===========================================================================
# All polygon files store coordinates as (lon, lat) with no header.
# Browns_line.txt and Halifax_line.txt are tab-separated; CSVs are comma-separated.
# sf requires the first and last row of each ring to be identical (closed ring).

read_poly <- function(path, sep = ",") {
  df  <- read.table(path, sep = sep, header = FALSE)
  mat <- as.matrix(df[, 1:2])
  colnames(mat) <- c("lon", "lat")
  if (!isTRUE(all.equal(mat[1, ], mat[nrow(mat), ]))) {
    mat <- rbind(mat, mat[1, ])
  }
  mat
}

make_sfc <- function(mat) {
  st_sfc(st_polygon(list(mat)), crs = 4326)
}

# ===========================================================================
# Load pre-clipped CINAR polygon coordinates
# ===========================================================================
# poly_*.csv files are produced by export_for_R.m, which clips each polygon
# against all higher-priority polygons (subtract() in MATLAB).
# Priority order: GMB150 > JB > GeorgesNEC > BOF > WSS > EGOM > Browns > Halifax
# No R-side clipping is needed — the files are already non-overlapping.
# GMB_200 and JB_250 are display-only and excluded from point assignment.

cat("Loading pre-clipped CINAR polygon coordinate files...\n")

cinar_polygons <- st_sf(
  CINAR_poly = c(7L, 3L, 6L, 8L, 1L, 2L, 4L, 5L),
  cinar_name = c("GMB150", "JB", "GeorgesNEC", "BOF", "WSS", "EGOM", "Browns", "Halifax"),
  geometry   = st_sfc(
    make_sfc(read_poly(file.path(work_dir, "data", "poly_GMB_150.csv")))[[1]],
    make_sfc(read_poly(file.path(work_dir, "data", "poly_JB_deep.csv")))[[1]],
    make_sfc(read_poly(file.path(work_dir, "data", "poly_GeorgesNEC.csv")))[[1]],
    make_sfc(read_poly(file.path(work_dir, "data", "poly_BOF_latlon.csv")))[[1]],
    make_sfc(read_poly(file.path(work_dir, "data", "poly_WSS_broad.csv")))[[1]],
    make_sfc(read_poly(file.path(work_dir, "data", "poly_EGOM_broad.csv")))[[1]],
    make_sfc(read_poly(file.path(work_dir, "data", "poly_Browns_line.csv")))[[1]],
    make_sfc(read_poly(file.path(work_dir, "data", "poly_Halifax_line.csv")))[[1]],
    crs = 4326
  )
)

cat(sprintf("CINAR polygon object: %d polygons\n", nrow(cinar_polygons)))

# ===========================================================================
# Load EcoMon strata
# ===========================================================================
# Stratum polygons come from EMstrata_v4_coords.csv, which was exported from
# EMstrata_v4.mat (MATLAB) using the snippet in run_pipeline.R.
# The sf object is cached as ne_strata_cache.rds after the first build.
# Delete ne_strata_cache.rds to force a rebuild from the CSV.

cat("Loading EcoMon strata from EMstrata_v4_coords.csv...\n")

# Reads polygon coordinates exported from EMstrata_v4.mat via the MATLAB snippet
# in the README. Cached as ne_strata_cache.rds after first build.
# Delete ne_strata_cache.rds to force a rebuild.
strata_cache <- file.path(work_dir, "cache", "ne_strata_cache.rds")

if (!file.exists(strata_cache)) {
  coords_file <- file.path(work_dir, "data", "EMstrata_v4_coords.csv")
  if (!file.exists(coords_file)) {
    stop(
      "EMstrata_v4_coords.csv not found.\n",
      "Generate it by running the MATLAB export snippet in the README:\n",
      "  load('EMstrata_v4.mat'); ... writetable(T, 'EMstrata_v4_coords.csv')"
    )
  }
  cat("  Building sf polygons from EMstrata_v4_coords.csv...\n")

  strata_coords <- read.csv(coords_file)

  # Build one sf polygon per stratum index (14-47)
  strata_list <- lapply(split(strata_coords, strata_coords$stratum_id), function(df) {
    mat <- as.matrix(df[, c("lon", "lat")])
    if (!isTRUE(all.equal(mat[1, ], mat[nrow(mat), ]))) mat <- rbind(mat, mat[1, ])
    st_polygon(list(mat))
  })

  ecomon_polygons <- st_sf(
    EcoMon_poly = as.integer(names(strata_list)),
    geometry    = st_sfc(strata_list, crs = 4326)
  )

  saveRDS(ecomon_polygons, strata_cache)
  cat("  Cached to:", strata_cache, "\n")
} else {
  ecomon_polygons <- readRDS(strata_cache)
  cat("  Loaded from local cache.\n")
}

cat(sprintf("Loaded %d EcoMon strata (ID range: %d - %d)\n",
            nrow(ecomon_polygons),
            min(ecomon_polygons$EcoMon_poly),
            max(ecomon_polygons$EcoMon_poly)))

# ===========================================================================
# Process each processed CSV file
# ===========================================================================

csv_files <- list.files(input_dir, pattern = "_processed\\.csv$", full.names = TRUE)
cat(sprintf("\nFound %d processed CSV files\n", length(csv_files)))

for (filepath in csv_files) {
  filename <- basename(filepath)
  cat(sprintf("\n%s\nProcessing: %s\n", strrep("=", 60), filename))

  tryCatch({

    data <- read.csv(filepath)
    cat(sprintf("Loaded: %d rows x %d cols\n", nrow(data), ncol(data)))

    # Check required columns
    missing <- setdiff(c("X", "Y", "REGION"), names(data))
    if (length(missing) > 0) {
      cat(sprintf("Warning: Missing columns: %s. Skipping.\n", paste(missing, collapse = ", ")))
      next
    }

    # Filter to desired regions
    data <- data[data$REGION %in% desired_regions, ]
    cat(sprintf("After region filter: %d rows\n", nrow(data)))
    if (nrow(data) == 0) next

    # Convert to sf points (X = lon, Y = lat; original columns retained)
    pts <- st_as_sf(data, coords = c("X", "Y"), crs = 4326, remove = FALSE)

    # ---- CINAR assignment ------------------------------------------------
    # Non-overlapping clipped geometry means each point matches at most one polygon.
    # Deduplicate by Label as a safeguard against boundary-point double-matches.
    cat("Assigning CINAR polygons...\n")
    pts <- suppressWarnings(st_join(pts, cinar_polygons[, c("CINAR_poly", "cinar_name")],
                                    join = st_within, left = TRUE))
    pts <- pts[!duplicated(pts$Label), ]
    pts$CINAR_poly[is.na(pts$CINAR_poly)] <- 0L
    pts$cinar_name <- NULL

    # ---- EcoMon assignment -----------------------------------------------
    cat("Assigning EcoMon polygons...\n")
    pts <- suppressWarnings(st_join(pts, ecomon_polygons[, "EcoMon_poly"],
                                    join = st_within, left = TRUE))
    pts <- pts[!duplicated(pts$Label), ]
    pts$EcoMon_poly[is.na(pts$EcoMon_poly)] <- 0L

    # Drop sf geometry, return to plain data frame
    result <- st_drop_geometry(pts)

    # ---- Write output ----------------------------------------------------
    base_name <- sub("_processed\\.csv$", "", filename)
    out_file  <- file.path(output_dir, paste0(base_name, "_processed_polygons.csv"))
    write.csv(result, out_file, row.names = FALSE)

    # ---- Summary ---------------------------------------------------------
    n_cinar_assigned   <- sum(result$CINAR_poly > 0)
    n_cinar_unassigned <- sum(result$CINAR_poly == 0)
    n_ecomon_assigned  <- sum(result$EcoMon_poly > 0)

    cat("CINAR polygon assignments:\n")
    for (id in sort(unique(result$CINAR_poly[result$CINAR_poly > 0]))) {
      nm  <- cinar_polygons$cinar_name[cinar_polygons$CINAR_poly == id]
      cnt <- sum(result$CINAR_poly == id)
      cat(sprintf("  %s (%d): %d points\n", nm, id, cnt))
    }
    cat(sprintf("  Unassigned: %d points\n", n_cinar_unassigned))
    cat(sprintf("EcoMon assignments: %d points assigned\n", n_ecomon_assigned))
    cat(sprintf("Saved: %s\n", basename(out_file)))

  }, error = function(e) {
    cat(sprintf("Error processing %s: %s\n", filename, e$message))
  })
}

cat(sprintf("\n%s\nProcessing complete!\n", strrep("=", 60)))
cat(sprintf("Output files saved to: %s\n", output_dir))
